# handle uncaught exceptions

How do tools like Sentry catch uncaught Python exceptions? 🐍

One key mechanism: they hook into `sys.excepthook` 💡

This built-in hook is triggered when an exception goes uncaught at the top level of your program. Sentry installs its own handler here to capture and send the error before your program exits. 🚀

Using Sentry is a smart default 😎, but if you want to do this manually, you can do so like this 📈

Let's be creative actually, here is an example of sending an email when the program crashes (requires Sendgrid lib + API key):

```
import sys
import traceback

import sendgrid
from sendgrid.helpers.mail import Mail

sg = sendgrid.SendGridAPIClient(api_key="YOUR_API_KEY")


def notify_exception(exc_type, exc_value, tb):
    details = "".join(traceback.format_exception(exc_type, exc_value, tb))
    msg = Mail(
        from_email="alerts@yourdomain.com",
        to_emails="you@example.com",
        subject="🚨 Python Exception",
        plain_text_content=details,
    )
    sg.send(msg)


sys.excepthook = notify_exception

# Unhandled crash should trigger the notification
1 / 0
```

#exceptions
